### PR TITLE
cmake: Build `bitcoin-qt` executable

### PR DESCRIPTION
**Pull request by hebasto**
A new configuration option `WITH_GUI` has been added. Its valid values are `AUTO`, `Qt5`, `OFF`.

### CODE REVIEW

**Comment by hebasto** [comment]
cc @fanquake @TheCharlatan @theuni @vasild @luke-jr @willcl-ark @stickies-v @pablomartin4btc @m3dwards

**Comment by hebasto** [comment]
The branch fails to link on macOS (

Working on it.

**Comment by hebasto** [comment]
> Also the GUI has been enabled in the CI (except for MSVC, for now).

Now, the GUI is enabled for the MSVC CI job as well.

Ready for review :)

File: depends/toolchain.cmake.in
Comment (fanquake) [inline depends/toolchain.cmake.in:95]: Can't remember the exact comments from yesterday, but I think this could use some docs to explain why the SDK is leaking out of the compiler invocation.

File: src/qt/winshutdownmonitor.h
Comment (hebasto) [reply src/qt/winshutdownmonitor.h:13]: A build error happens only when vcpkg's Qt packages are used. It is a not CMake-specific issue, rather vcpkg-specific.

> If this change is needed shouldn't it already be needed in master...?

According to the provided [explanation](https://stackoverflow.com/questions/4845198/fatal-error-no-target-architecture-in-visual-studio), the change is correct. However, I was a bit vague about whether the motivation is compelling enough for the master branch in the non-CMake context.

File: depends/toolchain.cmake.in
Comment (hebasto) [reply depends/toolchain.cmake.in:87]: I've added an explanation comment: https://github.com/hebasto/bitcoin/blob/b0a1bae9f81ff59ca63915b293e9202a9609a4e8/depends/toolchain.cmake.in#L95-L99

File: depends/toolchain.cmake.in
Comment (hebasto) [reply depends/toolchain.cmake.in:99]: I've added an explanation comment: https://github.com/hebasto/bitcoin/blob/b0a1bae9f81ff59ca63915b293e9202a9609a4e8/depends/toolchain.cmake.in#L95-L99

**Comment by m3dwards** [comment]
I've successfully built and ran bitcoin-qt on Mac (aarch64) and Debian 12 (x86) using system libraries. 

I've been trying to build on Windows but the vcpkg is so large I keep running out of hard disk space on my small laptop. Will update here if I get it built on Windows.

File: depends/toolchain.cmake.in
Comment (hebasto) [reply depends/toolchain.cmake.in]: De-globalized. An explanation comment added: https://github.com/hebasto/bitcoin/blob/c6c2cf964abf01b0cb21e4c6fcf90fd18ca72b2e/src/qt/CMakeLists.txt#L18-L26

**Comment by m3dwards** [comment]
@hebasto I did get a Windows build completed using your tip of not doing debug vcpkg builds.

I don't know if this is expected or not but when I run bitcoin-qt I also get a blank cmd prompt run which if I close it closes the qt GUI. Expected?

File: src/qt/winshutdownmonitor.h
Comment (hebasto) [reply src/qt/winshutdownmonitor.h:13]: > If this change is needed shouldn't it already be needed in master...?

Proposed in https://github.com/bitcoin-core/gui/pull/789.

File: depends/toolchain.cmake.in
Comment (pablomartin4btc) [reply depends/toolchain.cmake.in]: I agree with this change, moving it to CMakeLists.txt to de-globalized, but it was copied and left in `toolchain.cmake.in`, is this correct for when the dependencies are built?

Also, `set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY BOTH)` is required not only for cross-compilation but also for Mac or Apple devices, as I found on "_Professional CMake - A Practical Guide by Craig Scott_" (2023):

_On Apple platforms, consider carefully before setting `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY` to `ONLY`, as libraries may be built as fat binaries which support multiple target platforms. These fat binaries may not reside under target platform-specific paths, so it may still be necessary to search host platform paths to find them._

Do we need `set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)` to change it to `BOTH` as well?

Comment (hebasto) [reply depends/toolchain.cmake.in]: > Do we need `set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)` to change it to `BOTH` as well?

I don't think so, because our both native macOS CI jobs are green. And we set the `CMAKE_FIND_ROOT_PATH` only for cross-compiling (please note that the `CMAKE_FIND_ROOT_PATH_MODE_*` variables values affect how the `CMAKE_FIND_ROOT_PATH` is applied during searching routines).

**Review by pablomartin4btc — COMMENTED** [review]
tACK c6c2cf964abf01b0cb21e4c6fcf90fd18ca72b2e

<details>
<summary>Tested on Ubuntu 22.04, ran all tests and executed <code>bitcoin-qt</code> successfully</summary>

image

image

image

image

</details>

Pending of testing it on cross-compilation for both `Windows` and `Android`.

**Comment by hebasto** [comment]
@pablomartin4btc 
> Pending of testing it on cross-compilation for both `Windows` and `Android`.

I don't think this branch supports cross-compilation for Android. Such a feature is planned for the later.

File: depends/toolchain.cmake.in
Comment (TheCharlatan) [reply depends/toolchain.cmake.in:99]: I'm still confused by this. Does this relate to the previous commit? If I unset `CMAKE_FRAMEWORK_PATH` and drop the previous commit, it still configures, but then fails to link bitcoin-qt.

Comment (hebasto) [reply depends/toolchain.cmake.in:99]: > I'm still confused by this.

What exactly is the source of your confusion: the code or the comment?

Comment (TheCharlatan) [reply depends/toolchain.cmake.in:99]: The code, but I guess I should just look at what cmake does here.

Comment (hebasto) [reply depends/toolchain.cmake.in:99]: https://cmake.org/cmake/help/latest/command/find_library.html describes the role of the `CMAKE_FRAMEWORK_PATH` variable.

Comment (hebasto) [reply depends/toolchain.cmake.in:99]: For Qt's part, please see https://github.com/qt/qtbase/blob/da6e958319e95fe564d3b30c931492dd666bfaff/mkspecs/features/data/cmake/Qt5BasicConfig.cmake.in#L77-L134

**Review by vasild — APPROVED** [review]
ACK c6c2cf964abf01b0cb21e4c6fcf90fd18ca72b2e

`bitcoin-qt` is build and works on FreeBSD 14 with clang 18.

However, some tweaks were needed to get the compilation to succeed (I guess not due to this PR, but due to the C++20 switch from some time ago?).

1. `clang-scan-deps` is not found. The default compiler `/usr/bin/cc` is clang 16 and I do not think `clang-scan-deps` is installed, or at least I could not find under what name, nor did cmake. Anyway, I prefer compiling with the latest compiler clang 18 which installs the tool under `/usr/local/bin/clang-scan-deps-devel`, thus I had to use `cmake ... -DCMAKE_CXX_COMPILER_CLANG_SCAN_DEPS:PATH=/usr/local/bin/clang-scan-deps-devel`. The error is:
```
        "CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND" -format=p1689 -- /usr/local/bin/clang++-devel -DSTD_ATOMIC_NEEDS_LINK_TO_LIBATOMIC  -std=c++20 -x c++ /tmp/b/CMakeFiles/CMakeScratch/TryCompile-VQkc4W/src.cxx -c -o CMakeFiles/cmTC_adea9.dir/src.cxx.o -MT CMakeFiles/cmTC_adea9.dir/src.cxx.o.ddi -MD -MF CMakeFiles/cmTC_adea9.dir/src.cxx.o.ddi.d > CMakeFiles/cmTC_adea9.dir/src.cxx.o.ddi.tmp && mv CMakeFiles/cmTC_adea9.dir/src.cxx.o.ddi.tmp CMakeFiles/cmTC_adea9.dir/src.cxx.o.ddi
        /bin/sh: CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND: not found
```

2. It cannot find `<cpuid.h>` included from `src/compat/cpuid.h`. That file is installed in `/usr/local/llvm-devel/lib/clang/18/include/cpuid.h`, thus `cmake ... -DCMAKE_CXX_FLAGS:STRING="-I/usr/local/llvm-devel/lib/clang/18/include"` was needed. I am not sure how this works with autotools or how it worked recently with cmake (I did not had to explicitly provide this include directory, maybe it is related to clang-scan-deps?).

File: depends/patches/qt/qt.pro
Comment (hebasto) [reply depends/patches/qt/qt.pro:8]: The "prl" files are not used by the current Autotools-based build system. Therefore, it was odd to install them just to remove them on the next step.

CMake, on the contrary, does use the "prl" files. It processes them to collect internal dependencies of a library, enabling it to locate them on the system.

File: depends/toolchain.cmake.in
Comment (hebasto) [reply depends/toolchain.cmake.in:99]: > Does this relate to the previous commit? If I unset `CMAKE_FRAMEWORK_PATH` and drop the previous commit, it still configures, but then fails to link bitcoin-qt.

No, it doesn't. The `Revert "build, qt: Do not install *.prl files"` commit does not actually made the "prl" files available, which is required for a proper configuration. This is done in the `cmake: Build `bitcoin-qt` executable` commit. The first commit is supposed to make further changes more meaningful and clear.

Unsetting `CMAKE_FRAMEWORK_PATH` would result in a build failure regardless of the commit.

File: depends/packages/qt.mk
Comment (hebasto) [reply depends/packages/qt.mk:290]: The goal of the `$(package)_stage_cmds` is to remove from the package everything unneeded, while keeping everything required.

The content of the `native/mkspecs/` and `lib/cmake/` directories is required for CMake configuration.

The `doc/` directory is hardly needed for the build process ;)

**Comment by hebasto** [comment]
> ... when I run bitcoin-qt I also get a blank cmd prompt run which if I close it closes the qt GUI.

Fixed.

**Review by vasild — APPROVED** [review]
ACK 71aa125fe0e0c24c1bcde0c1617ef8a0a8cd9949

**Comment by theuni** [comment]
Before reviewing today I've been doing compile tests for various platforms. Working so far for:
linux-x86_64 target on linux-x86_64 builder
mingw-w64 on linux-x86_64 builder
macos on linux-x86_64 builder
native macos

In order to build for native macos with no homebrew, I used the following branch to remove some cmake/pkg-config dependencies: https://github.com/theuni/bitcoin/commits/depends-less-pkgconfig/

A few of those patches have been proposed in other PRs, but not merged due to parity issues with their autotools equivalents, but they were good enough for testing this branch.

Will continue with actual review tomorrow.

**Review by pablomartin4btc — COMMENTED** [review]
tACK 71aa125fe0e0c24c1bcde0c1617ef8a0a8cd9949

<details>
<summary>I can't make it work on cross-building for Windows, I keep getting missing references. Tried WSL with Ubuntu 22.04 ad Ubuntu 22.04 itself.</summary>

```
make -C depends -j $(nproc) HOST=x86_64-w64-mingw32
mkdir build
cd build
cmake -S .. --toolchain ../depends/x86_64-w64-mingw32/share/toolchain.cmake
cmake --build . -j $(nproc)
ctest -j $(nproc)


```

e.g.:

```
[ 36%] Linking CXX static library libleveldb.a
CMakeFiles/bitcoin-tx.dir/objects.a(bitcoin-tx.cpp.obj): in function `void std::call_once<void (&)()>(std::once_flag&, void (&)())':
/usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/mutex:724: undefined reference to `std::__get_once_mutex()'
/usr/bin/x86_64-w64-mingw32-ld: /usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/mutex:726: undefined reference to `std::__set_once_functor_lock_ptr(std::unique_lock<std::mutex>*)'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoin-tx.dir/objects.a(bitcoin-tx.cpp.obj):/usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/x86_64-w64-mingw32/bits/gthr-default.h:700: undefined reference to `__once_proxy'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoin-tx.dir/objects.a(bitcoin-tx.cpp.obj): in function `void std::call_once<void (&)()>(std::once_flag&, void (&)())':
/usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/mutex:733: undefined reference to `std::__set_once_functor_lock_ptr(std::unique_lock<std::mutex>*)'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoin-tx.dir/objects.a(bitcoin-tx.cpp.obj):bitcoin-tx.cpp:(.rdata$.refptr.__once_proxy[.refptr.__once_proxy]+0x0): undefined reference to `__once_proxy'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoin-tx.dir/objects.a(bitcoin-tx.cpp.obj):bitcoin-tx.cpp:(.rdata$.refptr._ZSt14__once_functor[.refptr._ZSt14__once_functor]+0x0): undefined reference to `std::__once_functor'
/usr/bin/x86_64-w64-mingw32-ld: [ 36%] Built target leveldb
libbitcoin_common.a(system.cpp.obj): in function `GetNumCores()':
/home/pablo/workspace/bitcoin-hebasto/src/common/system.cpp:101: undefined reference to `std::thread::hardware_concurrency()'
[ 36%] Built target bitcoin_wallet
collect2: error: ld returned 1 exit status
gmake[2]: *** [src/CMakeFiles/bitcoin-tx.dir/build.make:107: src/bitcoin-tx.exe] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1575: src/CMakeFiles/bitcoin-tx.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2

```

</details>

I can cross-build successfully with previous PRs with `NO_QT=1` (e.g.: #31, #75).

```
clang --version
Ubuntu clang version 14.0.0-1ubuntu1.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

`mingw-w64 is already the newest version (8.0.0-1).`

`gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0`

`GNU Make 4.3`

`cmake version 3.22.1`

**Review by pablomartin4btc — COMMENTED** [review]
> > I can't make it work on cross-building for Windows, I keep getting missing references.
> 
> Could you make another try in the clean Docker container?

<details>
<summary>That works just fine, thanks!</summary>

```
~/workspace/bitcoin-hebasto-docker$ docker build -t bitcoin-builder .
[+] Building 1375.0s (13/13) FINISHED                                                                     docker:default
 => [internal] load build definition from Dockerfile                                                                0.0s
 => => transferring dockerfile: 788B                                                                                0.0s
 => [internal] load metadata for docker.io/library/ubuntu:22.04                                                     5.0s
 => [internal] load .dockerignore                                                                                   0.1s
 => => transferring context: 2B                                                                                     0.0s
 => [1/9] FROM docker.io/library/ubuntu:22.04@sha256:e6173d4dc55e76b87c4af8db8821b1feae4146dd47341e4d431118c7dd060  8.7s
 => => resolve docker.io/library/ubuntu:22.04@sha256:e6173d4dc55e76b87c4af8db8821b1feae4146dd47341e4d431118c7dd060  0.0s
 => => sha256:e6173d4dc55e76b87c4af8db8821b1feae4146dd47341e4d431118c7dd060a74 1.13kB / 1.13kB                      0.0s
 => => sha256:cb2af41f42b9c9bc9bcdc7cf1735e3c4b3d95b2137be86fd940373471a34c8b0 424B / 424B                          0.0s
 => => sha256:e34e831650c1bb0be9b6f61c6755749cb8ea2053ba91c6cda27fded9e089811f 2.30kB / 2.30kB                      0.0s
 => => sha256:29202e855b2021a2d7f92800619ed5f5e8ac402e267cfbb3d29a791feb13c1ee 29.55MB / 29.55MB                    7.9s
 => => extracting sha256:29202e855b2021a2d7f92800619ed5f5e8ac402e267cfbb3d29a791feb13c1ee                           0.6s
 => [2/9] RUN apt-get update &&     apt-get install -y --no-install-recommends     git ca-certificates cmake curl  70.0s
 => [3/9] RUN git clone --depth=1 --branch 240117-cmake-S https://github.com/hebasto/bitcoin.git /bitcoin          16.0s 
 => [4/9] WORKDIR /bitcoin                                                                                          0.1s 
 => [5/9] RUN make -C depends -j $(nproc) HOST=x86_64-w64-mingw32                                                 696.8s 
 => [6/9] RUN mkdir build                                                                                           0.3s 
 => [7/9] WORKDIR /bitcoin/build                                                                                    0.1s 
 => [8/9] RUN cmake -S .. --toolchain ../depends/x86_64-w64-mingw32/share/toolchain.cmake                          43.6s 
 => [9/9] RUN cmake --build . -j $(nproc)                                                                         524.5s 
 => exporting to image                                                                                              9.7s 
 => => exporting layers                                                                                             9.7s 
 => => writing image sha256:0af226fe285bd8430937bd657f4a059b231ddb0f2a7355040a5256171292e4c3                        0.0s 
 => => naming to docker.io/library/bitcoin-builder
```

</details>

I'll give it a go again tmw on WSL as well.

**Review by pablomartin4btc — APPROVED** [review]
> I'll give it a go again tmw on WSL as well.

tACK 71aa125fe0e0c24c1bcde0c1617ef8a0a8cd9949

Tested cross-build for win on WSL Ubuntu 22.04 usind Docker and running generated `bitcoin-qt.exe` on Windows successfully.

File: src/qt/CMakeLists.txt
Comment (hebasto) [reply src/qt/CMakeLists.txt:37]: You are right. I would consider this approach as temporary one until the https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/main/update-translations.py script is updated. Specifically, its `update_build_systems` function.

Also the `bitcoin_locale.qrc` file could be a build artifact rather than a source tree file.

Comment (TheCharlatan) [reply src/qt/CMakeLists.txt:37]: Maybe add a comment to make this clear?

Comment (hebasto) [reply src/qt/CMakeLists.txt:40]: > Maybe add a comment to make this clear?

Added.

File: depends/toolchain.cmake.in
Comment (theuni) [reply depends/toolchain.cmake.in]: I spent most of yesterday poking at this. It's extremely annoying. I don't understand why the platform libs are lumped in here.

Is QT the only offender? Do other cmake projects also search for their platform dependencies like this? Afaik it makes `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY` completely useless.

If qt is the only lib with this behavior, I'm wondering if we should patch around it and/or upstream a change?

Building with `-DQt5_EXCLUDE_STATIC_DEPENDENCIES:BOOL=ON` _almost_ fixes the problem by letting us link the required libs ourselves. Except that the static plugins don't honor it :\

Comment (hebasto) [reply depends/toolchain.cmake.in]: > Is QT the only offender?

It seems so.

Comment (hebasto) [reply depends/toolchain.cmake.in]: > If qt is the only lib with this behavior, I'm wondering if we should patch around it and/or upstream a change?

The Qt active development happens for version 6 now. Suggesting to postpone this decision until we also bump our Qt package to the next major version. I'm assuming that things might change to better in there.

Comment (theuni) [reply depends/toolchain.cmake.in]: Heh, we've been burned too many times by qt in the past to assume that it will actually become more reasonable in the future. It's more likely that they'll invent yet another wonky buildsystem :p

As far as this goes though, does the fact that this lives in `src/qt/CMakeLists.txt` mean that the `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY` gets scoped to QT only? Or does that apply the option globally?

File: src/qt/winshutdownmonitor.h
Comment (theuni) [reply src/qt/winshutdownmonitor.h]: Could you explain what's different about the vcpkg qt? Does that imply that the qt headers installed by depends include more windows headers than vcpkg's? That's the only way I could see this making a difference.

Again, I'm not arguing about correctness here, this change seems fine to me. But I'm trying to understand why it's necessary in one place and not another, and if that means that our build (or vcpkg's) is suboptimal.

File: depends/packages/qt.mk
Comment (theuni) [inline depends/packages/qt.mk:284]: Can you explain this change? It's been ages since I've messed with the qt build, but IIRC this was done to keep the installed libs to a minimum. Now we just install everything?

Comment (hebasto) [reply depends/packages/qt.mk:284]: Otherwise, the CMake-driven configuration fails. I cannot point at the exact error as I did that change a year ago and just don't remember the exact problem.

File: src/qt/winshutdownmonitor.h
Comment (hebasto) [reply src/qt/winshutdownmonitor.h]: > Could you explain what's different about the vcpkg qt?

Not exactly (

> Does that imply that the qt headers installed by depends include more windows headers than vcpkg's? That's the only way I could see this making a difference.

Could it be related to pre-compiled headers as well?

I don't think that the comparison to our depends is correct because the minimum code example (https://github.com/bitcoin-core/gui/pull/789#issuecomment-1904455841) cross compiles fine, but fails with MSVC.

File: depends/toolchain.cmake.in
Comment (hebasto) [reply depends/toolchain.cmake.in]: > does the fact that this lives in `src/qt/CMakeLists.txt` mean that the `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY` gets scoped to QT only?

Yes, it does.

Comment (theuni) [reply depends/toolchain.cmake.in]: Ok, great. In that case I care much less about this and am ok with punting.

File: src/qt/winshutdownmonitor.h
Comment (theuni) [reply src/qt/winshutdownmonitor.h]: Yes, PCH could definitely affect the include order and it's on for mingw. So that's at least a plausible explanation. I'll do a quick check with it disabled.

File: depends/packages/qt.mk
Comment (theuni) [reply depends/packages/qt.mk:284]: Ok. Sounds like maybe it just added some additional cmake targets added to our list? Otherwise I believe this means `$(package)_qt_libs` is now unused.

 Mind adding a comment?

"TODO: Investigate whether specific targets can be used here to minimize the amount of files/components installed.".

File: src/qt/winshutdownmonitor.h
Comment (theuni) [reply src/qt/winshutdownmonitor.h]: Actually, a quick test shows that the pch option only affects the qt build itself. So I doubt that's the culprit.

This isn't interesting enough to hold up merge.

**Review by theuni — APPROVED** [review]
ACK (after adding the depends comment) to keep things moving.

File: depends/packages/qt.mk
Comment (hebasto) [reply depends/packages/qt.mk:285]: > Mind adding a comment?
> 
> "TODO: Investigate whether specific targets can be used here to minimize the amount of files/components installed.".

Added.

**Comment by hebasto** [comment]
> ACK (after adding the depends comment) to keep things moving.

The comment has been added.

While approving this PR, do you mind to approve https://github.com/bitcoin-core/gui/pull/789 as well?

**Comment by TheCharlatan** [comment]
Post-merge ACK

**Review by pablomartin4btc — COMMENTED** [review]
post-merge ACK 7f4babc752c6dc2027020e68891b6546e0fe8178

After investigating and debugging `cmake` building process with @hebasto's help, failures reported on [previous comment](https://github.com/hebasto/bitcoin/pull/77#pullrequestreview-1837779126) were not related to the QT module. Root cause of that problem is still under investigation.

**Review by pablomartin4btc — COMMENTED** [review]
> After investigating and debugging `cmake` building process with @hebasto's help, failures reported on [previous comment](https://github.com/hebasto/bitcoin/pull/77#pullrequestreview-1837779126) were not related to the QT module. Root cause of that problem is still under investigation.

Just for reference the issue was inconsistency in the binary building due to the linker using incorrect library path, so first thing is that I was having installed both packages `g++-mingw-w64-x86-64` and `g++-mingw-w64-x86-64-win32`, as per our [doc](https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md#building-for-64-bit-windows) we only require `posix` version for win cross-compilation (I've removed the win32 version and I managed to build succesfully), digging on where the incorrect path was determined is that @hebasto found a second issue that's related with `cmake` itself and the way the implicit directories are set, so he [raised it](https://gitlab.kitware.com/cmake/cmake/-/issues/25670) in cmake gitlab repo.

**Pull request merged**
